shared/src/main/scala/mlscript/codegen/typescript/TsTypegen.scala

```diff
@@ -147,15 +147,49 @@ final class TsTypegenCodeBuilder {
     methodSourceCode
   }
 
+  def addTypegenTraitDef(traitInfo: TraitSymbol, methods: List[(String, Type)]): Unit = {
+    val traitName = traitInfo.lexicalName
+    val traitBody = traitInfo.body
+    val typeParams = traitInfo.params.iterator.map(SourceCode(_)).toList
+
+    // extend super traits
+    var traitDeclaration = SourceCode(s"export interface $traitName") ++ SourceCode.paramList(typeParams)
+    val superTraits = typeScope.resolveImplementedTraits(traitBody)
+    if (!superTraits.isEmpty) {
+      val superTraitsCode = superTraits.map(symb => {
+        val traitName = symb.lexicalName
+        val traitParams = symb.params.iterator.map(SourceCode(_)).toList
+        SourceCode(s"$traitName") ++ SourceCode.paramList(traitParams)
+      }).toList
+      traitDeclaration ++= SourceCode(" extends ") ++ SourceCode.sepBy(superTraitsCode)
+    }
+    traitDeclaration ++= SourceCode.space ++ SourceCode.openCurlyBrace
+
+    // add fields defined in the trait
+    traitBody.collectBodyFieldsAndTypes
+      .foreach{case (fieldVar, fieldType) => 
+        val fieldCode = toTsType(fieldType)(TypegenContext(fieldType), Some(true))
+        traitDeclaration += SourceCode("    ") ++ SourceCode(fieldVar.name) ++ SourceCode.colon ++ fieldCode
+      }
+
+    // add methods
+    methods.foreach { case (name, methodType) =>
+      traitDeclaration += addMethodDeclaration(traitInfo.params.toSet, name, methodType)
+    }
+
+    typegenCode += traitDeclaration
+  }
+
   def addTypeGenClassDef(classInfo: ClassSymbol, methods: List[(String, Type)]): Unit = {
     val className = classInfo.lexicalName
     val classBody = classInfo.body
     val baseClass = typeScope.resolveBaseClass(classBody)
     val typeParams = classInfo.params.iterator.map(SourceCode(_)).toList
-    var classSourceCode = SourceCode.empty
 
     // create mapping for class body fields and types
-    val bodyFieldAndTypes = getClassFieldAndTypes(classInfo)
+    // class body includes fields from all implemented traits
```

Comment:
Note: I don't think it's necessary (or desired) to include the fields inherited from other traits in the class declaration. Why do that? The only place where we need the fields is in the ctor parameter type, I think.

Reply:
TS interface behaves like a type, so the class implementing a trait must have the trait fields as well. This is the example I tried.

```ts
interface lol { x: number }
interface lolol { y: number }
class lololol implements lol, lolol {
    x: number
    y: number
    constructor(x: number, y: number) {
        this.x = x
        this.y = y
    }
}
```

Removing x and y declaration will cause ts to complain.

Reply:
Hah, interesting. I didn't know that about TS!